src/bls12_381/ec.rs

```diff
@@ -130,18 +136,104 @@ macro_rules! curve_impl {
+            /// x_2 = (-1 - sqrt(-3))/2 + (sqrt(-3) * t^2)/(1 + b + t^2)
+            /// x_3 = 1 - (1 + b + t^2)^2/(3 * t^2)
+            ///
+            /// Given t != 0 and t != 1 + b + t^2 != 0, at least one of
+            /// these three points (x1, x2, x3) is valid on the curve.
+            ///
+            /// In the paper, 1 + b + t^2 != 0 has no solutions, but for
+            /// E(Fq) in our construction, it does have two solutions.
+            /// We follow the convention of the paper by mapping these
+            /// to some arbitrary points; in our case, the positive/negative
+            /// fixed generator (with the parity of the y-coordinate
+            /// corresponding to the t value).
+            ///
+            /// Unlike the paper, which maps t = 0 to an arbitrary point,
+            /// we map it to the point at infinity. This arrangement allows
+            /// us to preserve sw_encode(t) = sw_encode(-t) for all t.
+            ///
+            /// We choose the smallest i such that x_i is on the curve.
+            /// We choose the corresponding y-coordinate with the same
+            /// parity, defined as the point being lexicographically larger
+            /// than its negative.
+            fn sw_encode(t: $basefield) -> Self {
+                // Handle the case t == 0
+                if t.is_zero() {
+                    return Self::zero()
+                }
+
+                // We choose the corresponding y-coordinate with the same parity as t.
+                let parity = t.parity();
+
+                // w = (t^2 + b + 1)^(-1) * sqrt(-3) * t
+                let mut w = t;
+                w.square();
+                w.add_assign(&$affine::get_coeff_b());
+                w.add_assign(&$basefield::one());
+
+                // Handle the case t^2 + b + 1 == 0
+                if w.is_zero()  {
+                    let mut ret = Self::one();
+                    if parity {
+                        ret.negate()
+                    }
+                    return ret
+                }
+
+                w = w.inverse().unwrap();
+                w.mul_assign(&$basefield::get_swenc_sqrt_neg_three());
+                w.mul_assign(&t);
+
+                // x1 = - wt  + (sqrt(-3) - 1) / 2
+                let mut x1 = w;
+                x1.mul_assign(&t);
+                x1.negate();
+                x1.add_assign(&$basefield::get_swenc_sqrt_neg_three_minus_one_div_two());
+                if let Some(p) = Self::get_point_from_x(x1, parity) {
+                    return p
+                }
+
+                // x2 = -1 - x1
+                let mut x2 = x1;
+                x2.negate();
+                x2.sub_assign(&$basefield::one());
+                if let Some(p) = Self::get_point_from_x(x2, parity) {
+                    return p
+                }
+
+                // x3 = 1/w^2 + 1
+                let mut x3 = w;
+                x3.square();
+                x3 = x3.inverse().unwrap();
```

Comment:
I suspect this inverse could be optimized away.

Reply:
`1/w^2` is `w^(q-4)`. Trouble is, this is much slower than egcd inverse.

Unless you have completely different formula for x3 in mind.

Reply:
* `A = t^2`
* `B = A + b + 1`
* `C = 3t^2`
* `D = B * C`
* `E = D^(-1)`
* `F = E * B` makes for `F = 1/(3t^2)`
* `G = E * C` makes for `G = 1/(t^2 + b + 1)`

I don't have time to write the rest of the formulas out, but I think that's one inversion that gets you everything you need.